### PR TITLE
feat: Probe use built-in, discarded healthcheck.sh

### PR DESCRIPTION
**Pull request by muicoder**
**Description**

Feature: #443 

**Type of change**

* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)
* Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Checklist**

- [x] Testing has been performed
- [x] No functionality is broken
- [ ] Documentation updated

### CODE REVIEW

**Comment by muicoder** [comment]
README.md
===
`https://github.com/redis/redis/blob/unstable/TLS.md`
1. Run `./utils/gen-test-certs.sh` to generate a root CA and a server
   certificate.
2. Run `ls tests/tls | xargs -I0`
```shell
ca.crt
ca.key
ca.txt
client.crt
client.key
openssl.cnf
redis.crt
redis.dh
redis.key
server.crt
server.key
```
3. Run `cd tests/tls` and `kubectl create secret generic redis-tls $(for i in *.{crt,key}; do echo "--from-file $i"; done | xargs)`, create Secret for Kubernetes
4. set CR.spec.TLS(standalone/cluster/replication/sentinel)
```yaml
spec:
  TLS:
    ca: ca.crt
    cert: redis.crt
    key: redis.key
    secret:
      secretName: redis-tls
```

**Comment by muicoder** [comment]
tested example, Here are some important fields.
===
```yaml
apiVersion: v1
kind: Pod
metadata:
  generateName: dev-standalone-
  labels:
    app: dev-standalone
    controller-revision-hash: dev-standalone-55cc76f7fc
    redis_setup_type: standalone
    role: standalone
    statefulset.kubernetes.io/pod-name: dev-standalone-0
  name: dev-standalone-0
  namespace: default
spec:
  containers:
  - env:
    - name: REDISCLI_AUTH
      valueFrom:
        secretKeyRef:
          key: password
          name: auth.dev-standalone
    - name: REDIS_TLS_CA_KEY
      value: /tls/ca.crt
    - name: REDIS_TLS_CERT
      value: /tls/redis.crt
    - name: REDIS_TLS_CERT_KEY
      value: /tls/redis.key
    - name: SERVER_MODE
      value: standalone
    - name: SETUP_MODE
      value: standalone
    - name: TLS_MODE
      value: "true"
    image: redis:6.2
    livenessProbe:
      exec:
        command:
        - redis-cli
        - -p
        - "6379"
        - --tls
        - --cert
        - /tls/client.crt
        - --key
        - /tls/client.key
        - --cacert
        - /tls/ca.crt
        - ping
    name: dev-standalone
    readinessProbe:
      exec:
        command:
        - redis-cli
        - -p
        - "6379"
        - --tls
        - --cert
        - /tls/client.crt
        - --key
        - /tls/client.key
        - --cacert
        - /tls/ca.crt
        - ping
    startupProbe:
      exec:
        command:
        - redis-cli
        - -p
        - "6379"
        - --tls
        - --cert
        - /tls/client.crt
        - --key
        - /tls/client.key
        - --cacert
        - /tls/ca.crt
        - ping
    volumeMounts:
    - mountPath: /tls
      name: tls-certs
      readOnly: true
    - mountPath: /etc/redis/external.conf.d
      name: external-config
    - mountPath: /usr/local/bin/docker-entrypoint.sh
      name: scripts
      subPath: entrypoint.sh
  - env:
    - name: REDIS_ADDR
      value: redis://localhost:6379
    - name: REDIS_EXPORTER_SKIP_TLS_VERIFICATION
      value: "true"
    - name: REDIS_EXPORTER_TLS_CA_CERT_FILE
      value: /tls/ca.crt
    - name: REDIS_EXPORTER_TLS_CLIENT_CERT_FILE
      value: /tls/client.crt
    - name: REDIS_EXPORTER_TLS_CLIENT_KEY_FILE
      value: /tls/client.key
    - name: REDIS_PASSWORD
      valueFrom:
        secretKeyRef:
          key: password
          name: auth.dev-standalone
    - name: SERVER_MODE
      value: standalone
    - name: SETUP_MODE
      value: standalone
    image: oliver006/redis_exporter:alpine
    name: redis-exporter
    ports:
    - containerPort: 9121
      name: redis-exporter
      protocol: TCP
    volumeMounts:
    - mountPath: /tls
      name: tls-certs
      readOnly: true
    - mountPath: /usr/local/bin/docker-entrypoint.sh
      name: scripts
      subPath: entrypoint.sh
  hostname: dev-standalone-0
  nodeName: sealos002
  nodeSelector:
    redis: ""
  securityContext:
    runAsNonRoot: false
  subdomain: dev-standalone-headless
  volumes:
  - configMap:
      defaultMode: 420
      name: config.dev-standalone
    name: external-config
  - configMap:
      defaultMode: 493
      name: config.dev-standalone
    name: scripts
  - name: tls-certs
    secret:
      defaultMode: 420
      secretName: redis-tls
```

**Comment by muicoder** [comment]
@shubham-cmyk Tested as expected, please check.

**Comment by muicoder** [comment]
> # tested TLS, Is #447 a bug? @shubham-cmyk
> https://github.com/redis/redis/blob/unstable/TLS.md

Not a bug, need to set subjectAltName when self-signing.
```shell
cat <<EOF >"openssl.conf"
[req]
req_extensions = v3_req
distinguished_name = req_distinguished_name
[req_distinguished_name]
[v3_req]
basicConstraints = CA:FALSE
keyUsage = nonRepudiation, digitalSignature, keyEncipherment
extendedKeyUsage = clientAuth, serverAuth
subjectAltName = @alt_names
[alt_names]
DNS.1 = $ReplicationSvc-*
DNS.2 = $ReplicationSvc
DNS.3 = $ReplicationSvc.$NAMESPACE.svc
EOF
openssl x509 -extensions v3_req -extfile "openssl.conf" ......
```

File: k8sutils/statefulset.go [hunk 1]
```diff
@@ -486,14 +519,28 @@ func getEnvironmentVariables(role string, enabledMetric bool, enabledPassword *b
...
-		Value: redisHost,
-	})
+	RedisPassword := "REDISCLI_AUTH"
+	if enabledMetric {
+		envVars = append(envVars, corev1.EnvVar{
+			Name: "REDIS_ADDR",
+			Value: func(role string, tls *redisv1beta1.TLSConfig) string {
+				prefix := "redis://localhost:"
+				if tls != nil {
+					prefix = "rediss://localhost:"
```
Comment (iamabhishek-dubey) [inline k8sutils/statefulset.go:529]: seems like a typo here

File: k8sutils/statefulset.go [hunk 2]
```diff
@@ -486,14 +519,28 @@ func getEnvironmentVariables(role string, enabledMetric bool, enabledPassword *b
...
+				}
+			}(role, tlsConfig),
+		})
+		RedisPassword = "REDIS_PASSWORD"
+	}
 
 	if enabledPassword != nil && *enabledPassword {
 		envVars = append(envVars, corev1.EnvVar{
-			Name: "REDIS_PASSWORD",
+			Name: RedisPassword,
```
Comment (iamabhishek-dubey) [inline k8sutils/statefulset.go:543]: What's the need of updating this?

**Review by iamabhishek-dubey — CHANGES_REQUESTED** [review]

File: k8sutils/statefulset.go (hunk 2 above)
Comment (muicoder) [reply k8sutils/statefulset.go:543]: redis-cli uses [REDISCLI_AUTH](https://github.com/redis/redis/blob/12826fa38f0c481cce36f30b96b82028b1def0cb/src/redis-cli.c) by default
`NOTE: For security reasons, provide the password to redis-cli automatically via the REDISCLI_AUTH environment variable.`
redis_exporter uses [REDIS_PASSWORD](https://github.com/oliver006/redis_exporter/blob/48b92fc8d37f83b70ded121fc78f833e904ea5c4/main.go#L59) by default.

File: k8sutils/statefulset.go (hunk 1 above)
Comment (muicoder) [reply k8sutils/statefulset.go]: Updated to `redisHost`

**Comment by muicoder** [comment]
@shubham-cmyk Can you merge?

**Comment by shubham-cmyk** [comment]
@muicoder I have to test it this would be merged in next release which is going to be soon.

File: k8sutils/statefulset.go [hunk 3]
```diff
@@ -511,7 +511,45 @@ func getVolumeMount(name string, persistenceEnabled *bool, externalConfig *strin
...
 
 // getProbeInfo generate probe for Redis StatefulSet
-func getProbeInfo(probe *redisv1beta1.Probe) *corev1.Probe {
+func getProbeInfo(params containerParameters, probeType string) *corev1.Probe {
+	probePort := redisPort
+	if params.Role == "sentinel" {
+		probePort = sentinelPort
+	}
+
+	ProbeCommand := []string{
```
Comment (genofire) [inline k8sutils/statefulset.go:520]: inside of a function a variable should start with small later in golang (it is not accesseble from outside)

Comment (muicoder) [reply k8sutils/statefulset.go]: Renamed

**Review by genofire — APPROVED** [review]

**Comment by genofire** [comment]
any update @muicoder ?

**Comment by muicoder** [comment]
> any update @muicoder ?

All as expected

**Comment by genofire** [comment]
Ups sorry, i mean @shubham-cmyk for merging

**Comment by shubham-cmyk** [comment]
To be added in v0.15.0

**Comment by muicoder** [comment]
This can also be merged.

**Comment by shubham-cmyk** [comment]
This is not complete PR! You are only passing fewer keys and cert 

```
if params.TLSConfig.CaKeyFile != "" {
			caCert = params.TLSConfig.CaKeyFile
		}
```


check  here https://redis.io/docs/management/security/encryption/ : 

```
./src/redis-cli --tls \
    --cert ./tests/tls/redis.crt \
    --key ./tests/tls/redis.key \
    --cacert ./tests/tls/ca.crt
```

**Comment by muicoder** [comment]
Configured a full TLS certificate.

**Comment by muicoder** [comment]
Improved the use of env to load TLS certificates.

**Comment by shubham-cmyk** [comment]
@muicoder I won't be merged in the v0.15.0 because it would also require the change in the redis image I would merge it after releasing the v0.15.0 this change would be visible from v0.16.0

**Comment by genofire** [comment]
any update?

**Comment by genofire** [comment]
is there a reason, why this was closed?

does any other patch get the exporter in combination with sentinal working?

**Comment by muicoder** [comment]
> is there a reason, why this was closed?
> 
> does any other patch get the exporter in combination with sentinal working?

Repush later based on master branch (conflicting)

**Comment by muicoder** [comment]
Can't seem to reopen it?

**Comment by wkd-woo** [comment]
any update?

**Comment by shubham-cmyk** [comment]
I would get this merged in this release

**Comment by drivebyer** [comment]
I noticed that it has been a long time since it was last updated, but I think this change is very important. I would like to raise another PR to implement it.

**Comment by muicoder** [comment]
> I noticed that it has been a long time since it was last updated, but I think this change is very important. I would like to raise another PR to implement it.

Welcome 👍🏻 😀 👍🏼 👍🏽

**Comment by drivebyer** [comment]
It's been done by https://github.com/OT-CONTAINER-KIT/redis-operator/pull/907, thanks @muicoder

**Pull request closed without merging**